File: python/components/openedheaterdw0.py
"""
 General Object-oriented Abstraction and JSON Textual Model of Rankine Cycle 

class Openedheaterdw0

                      ↓   inPort extracted steam
                  ┌───┴───┐
 feedwater        │       │
 outPort_fw     ← ┤       │← inPort_fw  feedwater
                  │       │
                  └───────┘

 json object example:

     {
            "name": "OpenedFeedwaterHeater1",
            "devtype": "FWH-OPEN-DW0",
            "iPort":i,
            "iPort_fw":j,
            "oPort_fw":k
     }

  Author:Cheng Maohua  Email: cmh@seu.edu.cn

"""
from .port import *

class OpenedheaterDw0:

    energy = 'internel'
    devtype = "FWH-OPEN-DW0"

    def __init__(self, dictDev):
        """
        Initializes the Opened feedwater with the conditions
        """
        self.name = dictDev['name']
        self.iPort = [Port(dictDev['iPort'])]
        self.iPort_fw = [Port(dictDev['iPort_fw'])]
        self.oPort_fw = [Port(dictDev['oPort_fw'])]

        # map the name of port to the port obj
        self.portdict = {
            "iPort": self.iPort,
            "iPort_fw": self.iPort_fw, 
            "oPort_fw": self.oPort_fw
        }

        self.heatAdded = 0
        self.heatExtracted = 0
        self.QExtracted = 0

    def state(self):
        pass

    def balance(self):
        """ mass and energy balance of the opened feedwater heater """
        # energy balance equation
        qes1 = self.iPort[0].h - self.oPort_fw[0].h
        qfw1 = self.oPort_fw[0].h - self.iPort_fw[0].h
        self.iPort[0].fdot = self.oPort_fw[0].fdot * qfw1/(qes1 + qfw1)
        # mass balance equation
        self.iPort_fw[0].fdot = self.oPort_fw[0].fdot - self.iPort[0].fdot

        self.heatAdded = self.iPort_fw[0].fdot * qfw1
        self.heatExtracted = self.iPort[0].fdot * qes1

    def sm_energy(self):
        ucovt = 3600.0 * 1000.0
        self.QExtracted = self.iPort[0].mdot * \
            (self.iPort[0].h - self.oPort_fw[0].h)/ucovt
        self.QAdded = self.iPort_fw[0].mdot * \
            (self.oPort_fw[0].h - self.iPort_fw[0].h)/ucovt

    def __str__(self):
        result = '\n' + self.name
        result += '\n' + " PORT " + Port.title
        result += '\n'+" iPort " + self.iPort[0].__str__()
        result += '\n'+" iPort_fw " + self.iPort_fw[0].__str__()
        result += '\n' +" oPort_fw "+ self.oPort_fw[0].__str__()

        result += '\nheatAdded(kJ/kg) \t{:>.2f}'.format(self.heatAdded)
        result += '\nheatExtracted(kJ/kg) \t{:>.2f}'.format(self.heatExtracted)
        result += '\nQAdded(MW) \t{:>.2f}'.format(self.QAdded)
        result += '\nQExtracted(MW)  \t{:>.2f}'.format(self.QExtracted)
        return result

File: cpp/src/openedheaterdw0.cpp
/* 

  OpenedHeaterDw0.cpp

*/

#include "openedheaterdw0.hpp"

OpenedHeaterDw0::OpenedHeaterDw0(umComponent dictComp)
{
    name = any_cast<const char *>(dictComp["name"]);
    iPort = new Port(any_cast<mPort>(dictComp["iPort"]));
    iPort_fw = new Port(any_cast<mPort>(dictComp["iPort_fw"]));
    oPort_fw = new Port(any_cast<mPort>(dictComp["oPort_fw"]));

    portdict = {{"iPort", iPort},
                {"iPort_fw", iPort_fw},
                {"oPort_fw", oPort_fw}};
    energy = "internel";
}

OpenedHeaterDw0::~OpenedHeaterDw0()
{
    delete iPort;
    delete iPort_fw;
    delete oPort_fw;
}

void OpenedHeaterDw0::state()
{
}

int OpenedHeaterDw0::balance()
{
    // energy balance equation
    double qes1 = iPort->h - oPort_fw->h;
    double qfw1 = oPort_fw->h - iPort_fw->h;
    iPort->fdot = oPort_fw->fdot * qfw1 / (qes1 + qfw1);
    //mass balance equation
    iPort_fw->fdot = oPort_fw->fdot - iPort->fdot;

    heatAdded = iPort_fw->fdot * qfw1;
    heatExtracted = iPort->fdot * qes1;
    if (isnan(heatAdded))
    {
        return 0;
    }
    else
        return 1;
}

void OpenedHeaterDw0::setportaddress()
{
    if (iPort != portdict["iPort"])
        iPort = portdict["iPort"];
    if (iPort_fw != portdict["iPort_fw"])
        iPort_fw = portdict["iPort_fw"];
    if (oPort_fw != portdict["oPort_fw"])
        oPort_fw = portdict["oPort_fw"];
}

string OpenedHeaterDw0::resultstring()
{
    string result;
    result = "\n" + name;
    result += "\n" + Port::title;
    result += "\n" + iPort->resultstring();
    result += "\n" + iPort_fw->resultstring();
    result += "\n" + oPort_fw->resultstring();
    result += "\nheatAdded(kJ/kg): " + to_string_with_precision<double>(heatAdded, 3);
    result += "\nheatExtracted(kJ/kg): " + to_string_with_precision<double>(heatExtracted, 3) + "\n";
    return result;
}